lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2902 @@
+from torch import Tensor, nn, einsum
+import torch.distributed as distributed
+from torch.optim import Optimizer
+from torch.cuda.amp import autocast
+
+from lerobot.common.policies.vqbet.configuration_vqbet import VQBeTConfig
+from huggingface_hub import PyTorchModelHubMixin
+from robomimic.models.base_nets import SpatialSoftmax
+
+from lerobot.common.policies.normalize import Normalize, Unnormalize
+from lerobot.common.policies.utils import get_device_from_parameters, populate_queues
+
+class VQBeTPolicy(nn.Module, PyTorchModelHubMixin):
+    """
+    VQ-BeT Policy as per "Behavior Generation with Latent Actions"
+    """
+
+    name = "vqbet"
+    def __init__(
+        self,
+        config: VQBeTConfig | None = None,
+        dataset_stats: dict[str, dict[str, Tensor]] | None = None,
+    ):
+        """
+        Args:
+            config: Policy configuration class instance or None, in which case the default instantiation of
+                the configuration class is used.
+            dataset_stats: Dataset statistics to be used for normalization. If not passed here, it is expected
+                that they will be passed with a call to `load_state_dict` before the policy is used.
+        """
+        super().__init__()
+        if config is None:
+            config = VQBeTConfig()
+        self.config = config
+        self.normalize_inputs = Normalize(
+            config.input_shapes, config.input_normalization_modes, dataset_stats
+        )
+        self.normalize_targets = Normalize(
+            config.output_shapes, config.output_normalization_modes, dataset_stats
+        )
+        self.unnormalize_outputs = Unnormalize(
+            config.output_shapes, config.output_normalization_modes, dataset_stats
+        )
+
+
+        # queues are populated during rollout of the policy, they contain the n latest observations and actions
+        self._obs_queues = None
+
+        self.vqbet = VQBeTModel(config)
+
+    def check_discretized(self):
+        return self.vqbet._action_head._vqvae_model.discretized
+
+
+    def reset(self):
+        """
+        Clear observation and action queues. Should be called on `env.reset()`
+        """
+        self._obs_queues = {
+            "observation.image": deque(maxlen=self.config.n_obs_steps),
+            "observation.state": deque(maxlen=self.config.n_obs_steps),
+        }
+        if self.config.n_action_steps is not None:
+            self._action_queue = deque([], maxlen=self.config.n_action_steps)
+
+    @torch.no_grad
+    def select_action(self, batch: dict[str, Tensor]) -> Tensor:
+        """Select a single action given environment observations.
+
+        This method wraps `select_actions` in order to return one action at a time for execution in the
+        environment. It works by managing the actions in a queue and only calling `select_actions` when the
+        queue is empty.
+        """
+
+        # seungjae TODO: implement averaging action over horizons
```

Comment:
I tested it, but seems similar or worse than just not averaging over the multiple action. (and even require more computation power, since it should predict action chunk every timestep) So determined to drop it.